pkg/alertmanager/amcfg.go

```diff
@@ -0,0 +1,494 @@
+	Alt  string `yaml:"alt,omitempty" json:"alt,omitempty"`
+	Href string `yaml:"href,omitempty" json:"href,omitempty"`
+}
+
+func loadCfg(s string) (*alertmanagerConfig, error) {
+	// Run upstream Load function to get any validation checks that it runs.
+	_, err := config.Load(s)
+	if err != nil {
+		return nil, err
+	}
+
+	cfg := &alertmanagerConfig{}
+	err = yaml.UnmarshalStrict([]byte(s), cfg)
+
+	return cfg, nil
+}
+
+func (c alertmanagerConfig) String() string {
+	b, err := yaml.Marshal(c)
+	if err != nil {
+		return fmt.Sprintf("<error creating config string: %s>", err)
+	}
+	return string(b)
+}
+
+type configGenerator struct {
+	logger  log.Logger
+	kclient kubernetes.Interface
+	store   *assetStore
+}
+
+func newConfigGenerator(logger log.Logger, kclient kubernetes.Interface, store *assetStore) *configGenerator {
+	if store == nil {
+		store = newAssetStore(kclient.CoreV1(), kclient.CoreV1())
+	}
+
+	cg := &configGenerator{
+		logger:  logger,
+		kclient: kclient,
+		store:   store,
+	}
+	return cg
+}
+
+func (cg *configGenerator) generateConfig(
+	ctx context.Context,
+	baseConfig alertmanagerConfig,
+	amConfigs map[string]*monitoringv1alpha1.AlertmanagerConfig,
+) ([]byte, error) {
+
+	// amConfigIdentifiers is a sorted slice of keys from
+	// amConfigs map, used to always generate the config in the
+	// same order
+	amConfigIdentifiers := make([]string, len(amConfigs))
+	i := 0
+	for k := range amConfigs {
+		amConfigIdentifiers[i] = k
+		i++
+	}
+	sort.Strings(amConfigIdentifiers)
+
+	subRoutes := []*route{}
+	for _, amConfigIdentifier := range amConfigIdentifiers {
+		crKey := types.NamespacedName{
+			Name:      amConfigs[amConfigIdentifier].Name,
+			Namespace: amConfigs[amConfigIdentifier].Namespace,
+		}
+
+		// add routes to subRoutes
+		if amConfigs[amConfigIdentifier].Spec.Route != nil {
+			subRoutes = append(subRoutes, convertRoute(amConfigs[amConfigIdentifier].Spec.Route, crKey, true))
+		}
+
+		// add receivers to baseConfig.Receivers
+		for _, receiver := range amConfigs[amConfigIdentifier].Spec.Receivers {
+			baseConfig.Receivers = append(baseConfig.Receivers, cg.convertReceiver(ctx, &receiver, crKey))
+		}
+
+		// add inhibitRules to baseConfig.InhibitRules
+		for _, inhibitRule := range amConfigs[amConfigIdentifier].Spec.InhibitRules {
+			baseConfig.InhibitRules = append(baseConfig.InhibitRules, convertInhibitRule(&inhibitRule, crKey))
+		}
+	}
+
+	// Append subroutes from base to the end, then replace with the new slice
+	subRoutes = append(subRoutes, baseConfig.Route.Routes...)
+	baseConfig.Route.Routes = subRoutes
+
+	return yaml.Marshal(baseConfig)
+}
+
+func convertRoute(in *monitoringv1alpha1.Route, crKey types.NamespacedName, firstLevelRoute bool) *route {
+
+	// Enforce continue to be true for main Route in a CR
+	cont := in.Continue
+	if firstLevelRoute {
+		cont = true
+	}
+
+	match := map[string]string{}
+	matchRE := map[string]string{}
+
+	for _, matcher := range in.Matchers {
+		if *matcher.Regex {
```

Comment:
need to check for `nil` or probably better to change `Regex` from `*bool` to `bool`.

Reply:
Changed to `bool`. The types in `alertmanager_config_types.go` are another thing that existed before I started working on this, so I didn't think about them too deeply. If there are any others you'd like me to change, let me know!